File: app/src/main/java/net/north101/android/ghplayertracker/CharacterFragment.java
package net.north101.android.ghplayertracker;

import android.graphics.drawable.ColorDrawable;
import android.os.Bundle;
import android.support.design.widget.Snackbar;
import android.support.v4.app.Fragment;
import android.support.v4.view.ViewCompat;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.DefaultItemAnimator;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.EditorInfo;
import android.widget.BaseAdapter;
import android.widget.EditText;
import android.widget.Spinner;
import android.widget.TextView;
import android.widget.Toast;

import net.north101.android.ghplayertracker.data.Character;
import net.north101.android.ghplayertracker.data.CharacterClass;
import net.north101.android.ghplayertracker.data.CharacterList;
import net.north101.android.ghplayertracker.data.CharacterPerk;
import net.north101.android.ghplayertracker.data.Level;
import net.north101.android.ghplayertracker.data.Perk;
import net.yslibrary.android.keyboardvisibilityevent.KeyboardVisibilityEvent;
import net.yslibrary.android.keyboardvisibilityevent.KeyboardVisibilityEventListener;
import net.yslibrary.android.keyboardvisibilityevent.Unregistrar;

import org.androidannotations.annotations.AfterViews;
import org.androidannotations.annotations.Background;
import org.androidannotations.annotations.EFragment;
import org.androidannotations.annotations.EditorAction;
import org.androidannotations.annotations.FocusChange;
import org.androidannotations.annotations.FragmentArg;
import org.androidannotations.annotations.InstanceState;
import org.androidannotations.annotations.ItemSelect;
import org.androidannotations.annotations.OptionsItem;
import org.androidannotations.annotations.OptionsMenu;
import org.androidannotations.annotations.ViewById;
import org.json.JSONException;
import org.json.JSONObject;
import org.parceler.Parcels;

import java.io.IOException;
import java.text.ParseException;
import java.util.ArrayList;
import java.util.List;

@EFragment(R.layout.character_layout)
@OptionsMenu(R.menu.character)
public class CharacterFragment extends Fragment {
    ActionBar actionBar;
    @ViewById(R.id.toolbar)
    Toolbar toolbar;
    @ViewById(R.id.max_health)
    TextView maxHealthView;
    @ViewById(R.id.xp_text)
    EditText xpTextView;
    @ViewById(R.id.levels_select)
    Spinner levelsView;
    @ViewById(R.id.gold_text)
    EditText goldTextView;
    @ViewById(R.id.perk_list)
    RecyclerView perkListView;
    @ViewById(R.id.perk_note_grid)
    RecyclerView perkNoteGridView;
    @ViewById(R.id.name)
    TextView nameView;
    @ViewById(R.id.minus_1_text)
    TextView minus1TextView;

    @InstanceState
    @FragmentArg("character")
    Character character;

    LevelAdapter levelAdapter;
    PerkAdapter perkAdapter;
    PerkNoteAdapter perkNoteAdapter;

    Unregistrar keyboardEeventListener;

    @AfterViews
    void afterViews() {
        ((AppCompatActivity) getActivity()).setSupportActionBar(toolbar);
        actionBar = ((AppCompatActivity) getActivity()).getSupportActionBar();

        CharacterClass characterClass = character.getCharacterClass();

        int iconId = Util.getImageResource(getContext(), "icon_" + characterClass.id);
        actionBar.setLogo(iconId);
        actionBar.setTitle(characterClass.name);
        actionBar.setBackgroundDrawable(new ColorDrawable(characterClass.color));

        List<CharacterPerk> characterPerkList = new ArrayList<>();
        for (int i = 0; i < character.getPerks().length; i++) {
            Perk perk = characterClass.perks.get(i);
            characterPerkList.add(new CharacterPerk(perk, character.getPerks()[i]));
        }

        LinearLayoutManager perkListLayoutManager = new LinearLayoutManager(getContext());
        perkListView.setLayoutManager(perkListLayoutManager);
        perkListView.setItemAnimator(new DefaultItemAnimator());
        perkAdapter = new PerkAdapter(characterPerkList);
        perkListView.setAdapter(perkAdapter);
        ViewCompat.setNestedScrollingEnabled(perkListView, false);

        GridLayoutManager perkNoteListLayoutManager = new GridLayoutManager(getContext(), 3);
        perkNoteGridView.setLayoutManager(perkNoteListLayoutManager);
        perkNoteGridView.setItemAnimator(new DefaultItemAnimator());
        perkNoteAdapter = new PerkNoteAdapter(character.getPerkNotes());
        perkNoteGridView.setAdapter(perkNoteAdapter);
        ViewCompat.setNestedScrollingEnabled(perkListView, false);

        levelAdapter = new LevelAdapter(characterClass.levels);
        levelsView.setAdapter(levelAdapter);

        nameView.setText(character.getName());
        updateXPText();
        updateLevelText();
        updateMinus1Text();
        updateGoldText();
        updateHealthText();
    }

    @Override
    public void onResume() {
        super.onResume();

        //HACK!
        keyboardEeventListener = KeyboardVisibilityEvent.registerEventListener(getActivity(), new KeyboardVisibilityEventListener() {
            @Override
            public void onVisibilityChanged(boolean isOpen) {
                if (!isOpen) {
                    AppCompatActivity activity = (AppCompatActivity) getActivity();
                    if (activity == null) return;

                    View view = activity.getCurrentFocus();
                    if (view instanceof EditText) {
                        view.clearFocus();
                        view.requestFocus();
                    }
                }
            }
        });
    }

    @Override
    public void onPause() {
        super.onPause();
        if (keyboardEeventListener != null) {
            keyboardEeventListener.unregister();
            keyboardEeventListener = null;
        }

        updateCharacter();
    }

    @OptionsItem(R.id.start)
    void onMenuStartClick() {
        updateCharacter();

        Bundle args = new Bundle();
        args.putParcelable("character", Parcels.wrap(character));

        Fragment fragment = new CharacterTrackerFragment_();
        fragment.setArguments(args);

        getActivity().getSupportFragmentManager().beginTransaction()
            .setCustomAnimations(R.anim.enter_from_right, R.anim.exit_to_left, R.anim.enter_from_left, R.anim.exit_to_right)
            .replace(R.id.content, fragment)
            .addToBackStack(null)
            .commit();
    }

    @ItemSelect(R.id.levels_select)
    void onLevelSelect(boolean selected, Level selectedItem) {
        character.setLevel(((Level) levelsView.getSelectedItem()).level);
        maxHealthView.setText(String.valueOf(selectedItem.health));
    }

    void updateHealthText() {
        Log.d("khhjv", String.valueOf(character.getLevel()));
        maxHealthView.setText(String.valueOf(character.getMaxHealth()));
    }

    void updateCharacter() {
        character.setName(nameView.getText().toString());
        parseXPText();
        parseGoldText();
        parseMinus1Text();

        for (int i = 0; i < perkAdapter.items.size(); i++) {
            CharacterPerk characterPerk = perkAdapter.items.get(i);
            character.getPerks()[i] = characterPerk.ticks;
        }
    }

    @EditorAction(R.id.xp_text)
    boolean onXPTextChange(TextView tv, int actionId) {
        if (actionId != EditorInfo.IME_ACTION_DONE) return false;

        parseXPText();
        return false;
    }

    @FocusChange(R.id.xp_text)
    void onXPTextFocus() {
        if (xpTextView == null || xpTextView.hasFocus()) return;

        parseXPText();
    }

    void parseXPText() {
        try {
            character.setXP(Integer.parseInt(xpTextView.getText().toString()));
        } catch (Exception e) {
            e.printStackTrace();
            Toast.makeText(getContext(), "Invalid XP value", Toast.LENGTH_SHORT).show();
        }
        updateXPText();
    }

    void updateXPText() {
        String text = String.valueOf(character.getXP());
        if (!text.equals(xpTextView.getText().toString())) {
            xpTextView.setText(text);
        }
    }

    void updateLevelText() {
        levelsView.setSelection(character.getLevel() - 1);
    }

    @EditorAction(R.id.gold_text)
    boolean onGoldTextChange(TextView tv, int actionId) {
        if (actionId != EditorInfo.IME_ACTION_DONE) return false;

        parseGoldText();
        return false;
    }

    @FocusChange(R.id.gold_text)
    void onGoldTextFocus() {
        if (goldTextView == null || goldTextView.hasFocus()) return;

        parseGoldText();
    }

    void parseGoldText() {
        try {
            character.setGold(Integer.parseInt(goldTextView.getText().toString()));
        } catch (Exception e) {
            e.printStackTrace();
            Toast.makeText(getContext(), "Invalid Gold value", Toast.LENGTH_SHORT).show();
        }
        updateGoldText();
    }

    void updateGoldText() {
        String text = String.valueOf(character.getGold());
        if (!text.equals(goldTextView.getText().toString())) {
            goldTextView.setText(text);
        }
    }

    @EditorAction(R.id.minus_1_text)
    boolean onMinus1TextChange(TextView tv, int actionId) {
        if (actionId != EditorInfo.IME_ACTION_DONE) return false;

        parseMinus1Text();
        return false;
    }

    @FocusChange(R.id.minus_1_text)
    void onMinus1TextFocus() {
        if (minus1TextView == null || minus1TextView.hasFocus()) return;

        parseMinus1Text();
    }

    void parseMinus1Text() {
        try {
            character.setMinus1(Integer.parseInt(minus1TextView.getText().toString()));
        } catch (Exception e) {
            e.printStackTrace();
            Toast.makeText(getContext(), "Invalid -1 Attack modifier value", Toast.LENGTH_SHORT).show();
        }
        updateMinus1Text();
    }

    void updateMinus1Text() {
        String text = String.valueOf(character.getMinus1());
        if (!text.equals(minus1TextView.getText().toString())) {
            minus1TextView.setText(text);
        }
    }

    @OptionsItem(R.id.save)
    void onMenuSaveClick() {
        saveCharacter();
    }

    void saveCharacter() {
        saveCharacterTask();
    }

    @Background
    void saveCharacterTask() {
        JSONObject data;
        try {
            data = CharacterList.loadJSON(getContext());
        } catch (IOException | JSONException | ParseException e) {
            e.printStackTrace();

            data = new JSONObject();
        }

        try {
            data.put(character.getId().toString(), character.toJSON());
            CharacterList.parse(getContext(), data).save(getContext());
        } catch (IOException | JSONException e) {
            e.printStackTrace();

            Snackbar.make(getView(), "Failed to save", Snackbar.LENGTH_SHORT).show();
            return;
        }

        Snackbar.make(getView(), "Saved", Snackbar.LENGTH_SHORT).show();
    }

    public class LevelAdapter extends BaseAdapter {
        public final Level[] items;

        public LevelAdapter(Level[] items) {
            this.items = items;
        }

        @Override
        public int getCount() {
            return this.items.length;
        }

        @Override
        public Level getItem(int i) {
            return items[i];
        }

        @Override
        public long getItemId(int i) {
            return i;
        }

        @Override
        public View getDropDownView(int position, View view, ViewGroup parent) {
            if (view == null) {
                view = LayoutInflater
                    .from(parent.getContext())
                    .inflate(android.R.layout.simple_dropdown_item_1line, parent, false);
            }
            ((TextView) view.findViewById(android.R.id.text1)).setText(String.valueOf(getItem(position).level));
            return view;
        }

        @Override
        public View getView(int position, View view, ViewGroup parent) {
            if (view == null) {
                view = LayoutInflater
                    .from(parent.getContext())
                    .inflate(R.layout.spinner_text_view, parent, false);
            }
            ((TextView) view).setText(String.valueOf(getItem(position).level));
            return view;
        }
    }
}

File: app/src/main/java/net/north101/android/ghplayertracker/CharacterFragment.kt
package net.north101.android.ghplayertracker

import android.arch.lifecycle.Observer
import android.arch.lifecycle.ViewModelProviders
import android.content.res.Configuration
import android.graphics.Color
import android.graphics.drawable.ColorDrawable
import android.os.Bundle
import android.support.design.widget.Snackbar
import android.support.v4.app.Fragment
import android.support.v7.app.ActionBar
import android.support.v7.app.AlertDialog
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.GridLayoutManager
import android.support.v7.widget.RecyclerView
import android.support.v7.widget.Toolbar
import net.north101.android.ghplayertracker.data.Character
import net.north101.android.ghplayertracker.data.CharacterData
import org.androidannotations.annotations.*
import org.json.JSONException
import org.json.JSONObject
import java.io.IOException
import java.text.ParseException

@OptionsMenu(R.menu.character)
@EFragment(R.layout.character_tracker_layout)
open class CharacterFragment : Fragment(), OnBackPressedListener {
    protected lateinit var actionBar: ActionBar
    @ViewById(R.id.toolbar)
    protected lateinit var toolbar: Toolbar
    @ViewById(R.id.list1)
    protected lateinit var listView1: RecyclerView
    @ViewById(R.id.list2)
    protected lateinit var listView2: RecyclerView

    protected lateinit var listAdapter1: CharacterAdapter
    protected lateinit var listAdapter2: CharacterAdapter

    @FragmentArg("character")
    @InstanceState
    protected lateinit var character: Character

    lateinit var classModel: ClassModel
    lateinit var characterModel: CharacterModel
    lateinit var trackerResultModel: TrackerResultModel

    override fun onCreate(state: Bundle?) {
        super.onCreate(state)

        classModel = ViewModelProviders.of(this.activity!!).get(ClassModel::class.java)
        characterModel = ViewModelProviders.of(this).get(CharacterModel::class.java)
        trackerResultModel = ViewModelProviders.of(this).get(TrackerResultModel::class.java)

        if (state == null) {
            characterModel.init(character)
        } else {
            characterModel.fromBundle(state.getBundle("character_model"))
        }
    }

    override fun onSaveInstanceState(state: Bundle) {
        state.putBundle("character_model", characterModel.toBundle())

        super.onSaveInstanceState(state)
    }

    @AfterViews
    fun afterViews() {
        if (trackerResultModel.gold > 0) {
            characterModel.character.gold.value += trackerResultModel.gold
            trackerResultModel.gold = 0
        }
        if (trackerResultModel.xp > 0) {
            characterModel.character.xp.value += trackerResultModel.xp
            trackerResultModel.xp = 0
        }

        (activity as AppCompatActivity).setSupportActionBar(toolbar)
        actionBar = (activity as AppCompatActivity).supportActionBar!!

        val characterClass = character.characterClass

        val iconId = Util.getImageResource(context!!, "icon_" + characterClass.id)
        actionBar.setLogo(iconId)
        actionBar.title = characterClass.name
        var c = characterClass.color
        val hsv = FloatArray(3)
        Color.colorToHSV(c, hsv)
        hsv[2] = hsv[2] * 2 / 3
        c = Color.HSVToColor(hsv)
        actionBar.setBackgroundDrawable(ColorDrawable(c))

        if (!this::listAdapter1.isInitialized) {
            listAdapter1 = CharacterAdapter()
        }
        listAdapter1.onNumberEditClick = callback@{
            val fragment = when (it) {
                "level" -> CharacterEditLevelDialog_.builder().build()
                "xp" -> CharacterEditXPDialog_.builder().build()
                "gold" -> CharacterEditGoldDialog_.builder().build()
                else -> null
            } ?: return@callback
            fragment.setTargetFragment(this@CharacterFragment, 0)
            fragment.show(fragmentManager, "CharacterNumberEdit")
        }
        listAdapter1.onItemAddClick = {
            val fragment = CharacterItemDialog_.builder().build()
            fragment.setTargetFragment(this, 0)

            fragment.show(fragmentManager, "CharacterItemDialog_")
        }
        listAdapter1.onItemEditClick = {
            val args = Bundle()
            args.putInt("index", characterModel.character.items.value.indexOf(it))

            val fragment = CharacterItemDialog_.builder().build()
            fragment.arguments = args
            fragment.setTargetFragment(this, 0)

            fragment.show(fragmentManager, "CharacterNoteDialog")
        }
        listAdapter1.onItemDeleteClick = {
            characterModel.character.items.value.remove(it)
            characterModel.character.items.value = characterModel.character.items.value

        }
        listAdapter1.onNoteAddClick = {
            val fragment = CharacterNoteDialog_.builder().build()
            fragment.setTargetFragment(this, 0)

            fragment.show(fragmentManager, "CharacterNoteDialog")
        }
        listAdapter1.onNoteEditClick = {
            val args = Bundle()
            args.putInt("index", characterModel.character.notes.value.indexOf(it.note))

            val fragment = CharacterNoteDialog_.builder().build()
            fragment.arguments = args
            fragment.setTargetFragment(this, 0)

            fragment.show(fragmentManager, "CharacterNoteDialog")
        }
        listAdapter1.onNoteDeleteClick = {
            characterModel.character.notes.value.remove(it.note)
            characterModel.character.notes.value = characterModel.character.notes.value
        }

        val listLayoutManager1 = GridLayoutManager(context, 3)
        listLayoutManager1.spanSizeLookup = object : GridLayoutManager.SpanSizeLookup() {
            override fun getSpanSize(position: Int): Int {
                val item = listAdapter1.getItem(position)
                return if (item is CharacterAdapter.PerkNote) {
                    1
                } else {
                    3
                }
            }
        }
        listView1.layoutManager = listLayoutManager1
        listView1.adapter = listAdapter1

        if (!this::listAdapter2.isInitialized) {
            listAdapter2 = CharacterAdapter()
        }

        val listLayoutManager2 = GridLayoutManager(context, 3)
        listLayoutManager2.spanSizeLookup = object : GridLayoutManager.SpanSizeLookup() {
            override fun getSpanSize(position: Int): Int {
                val item = listAdapter2.getItem(position)
                return if (item is CharacterAdapter.PerkNote) {
                    1
                } else {
                    3
                }
            }
        }
        listView2.layoutManager = listLayoutManager2
        listView2.adapter = listAdapter2

        if (resources.configuration.orientation == Configuration.ORIENTATION_PORTRAIT) {
            listAdapter1.display = CharacterAdapter.DisplayItems.Both
            listAdapter2.display = CharacterAdapter.DisplayItems.None
        } else {
            listAdapter1.display = CharacterAdapter.DisplayItems.Left
            listAdapter2.display = CharacterAdapter.DisplayItems.Right
        }
        listAdapter1.updateItems(characterModel.character)
        listAdapter2.updateItems(characterModel.character)

        characterModel.character.items.observe(this, Observer {
            listAdapter1.updateItems(characterModel.character)
            listAdapter2.updateItems(characterModel.character)
        })
        characterModel.character.notes.observe(this, Observer {
            listAdapter1.updateItems(characterModel.character)
            listAdapter2.updateItems(characterModel.character)
        })
    }

    override fun onBackPressed(): Boolean {
        if (characterModel.character.toParcel() == character)
            return false

        val builder = AlertDialog.Builder(context!!)
        builder.setTitle("Would you like to save your changes?")
            .setPositiveButton("Yes") { dialog, which ->
                saveCharacter(characterModel.character.toParcel(), Runnable {
                    fragmentManager!!.popBackStack()
                })
            }
            .setNegativeButton("No") { dialog, which ->
                fragmentManager!!.popBackStack()
            }
            .setNeutralButton("Cancel") { dialog, which ->
                dialog.dismiss()
            }
            .show()

        return true
    }

    @OptionsItem(R.id.start)
    fun onMenuStartClick() {
        val args = Bundle()
        args.putParcelable("character", characterModel.character.toParcel())

        val fragment = TrackerFragment_()
        fragment.arguments = args
        fragment.setTargetFragment(this, 1)

        fragmentManager!!.beginTransaction()
            .replace(R.id.content, fragment)
            .addToBackStack(null)
            .commit()
    }

    @OptionsItem(R.id.save)
    fun onMenuSaveClick() {
        character = characterModel.character.toParcel()
        saveCharacter(character, null)
    }

    fun saveCharacter(character: Character, callback: Runnable?) {
        saveCharacterTask(character, callback)
    }

    open fun saveCharacterTask(character: Character, callback: Runnable?) {
        val data = try {
            CharacterData.load(context!!)
        } catch (e: IOException) {
            e.printStackTrace()
            null
        } catch (e: JSONException) {
            e.printStackTrace()
            null
        } catch (e: ParseException) {
            e.printStackTrace()
            null
        } ?: CharacterData(context!!, JSONObject())

        data.update(character)

        try {
            data.save()

            classModel.characterList.load()
        } catch (e: IOException) {
            e.printStackTrace()

            Snackbar.make(activity!!.findViewById(R.id.content), "Failed to save", Snackbar.LENGTH_SHORT).show()
            return
        } catch (e: JSONException) {
            e.printStackTrace()
            Snackbar.make(activity!!.findViewById(R.id.content), "Failed to save", Snackbar.LENGTH_SHORT).show()
            return
        }

        Snackbar.make(activity!!.findViewById(R.id.content), "Saved", Snackbar.LENGTH_SHORT).show()

        if (callback != null) {
            activity!!.runOnUiThread(callback)
        }
    }
}


@EBean
abstract class CharacterNumberDialog : EditNumberDialog() {
    lateinit var characterModel: CharacterModel

    @AfterViews
    override fun afterViews() {
        characterModel = ViewModelProviders.of(this.targetFragment!!).get(CharacterModel::class.java)

        super.afterViews()
    }
}


@EFragment
open class CharacterEditLevelDialog : CharacterNumberDialog() {
    override val title = "Edit Level"

    override var value: Int
        get() = characterModel.character.level.value
        set(value) {
            characterModel.character.level.value = value
        }
}


@EFragment
open class CharacterEditXPDialog : CharacterNumberDialog() {
    override val title = "Edit XP"

    override var value: Int
        get() = characterModel.character.xp.value
        set(value) {
            characterModel.character.xp.value = value
        }
}


@EFragment
open class CharacterEditGoldDialog : CharacterNumberDialog() {
    override val title = "Edit Gold"

    override var value: Int
        get() = characterModel.character.gold.value
        set(value) {
            characterModel.character.gold.value = value
        }
}
